compiler/hash-tree-def/src/emit.rs

```diff
@@ -760,19 +757,113 @@ fn emit_walker(
+
+    // We have one macro case prefixed with `@@` for each node type.
+    let default_impl_macro_cases = tree_def.nodes.keys().map(|node_name| {
+        let node_ret = format_ident!("{}Ret", node_name.to_string().to_case(Case::Pascal));
+        let node_name = format_ident!("{}", node_name.to_string().to_case(Case::Pascal));
+        let visit_node = format_ident!("visit_{}", node_name.to_string().to_case(Case::Snake));
+        let self_param = if self_mut { quote!(&mut self) } else { quote!(&self) };
+        let walk_node_fn_name =
+            format_ident!("walk_{}", node_name.to_string().to_case(Case::Snake),);
+
+        let impl_body = if is_leaf_node(&node_name, tree_def) {
+            quote! {
+                Ok(())
+            }
+        } else {
+            quote! {
+                let _ = #root_module::#walk_mod_name::#walk_node_fn_name(self, node)?;
+                Ok(())
+            }
+        };
+
+        quote! {
+            (@@ #node_name) => {
+                type #node_ret = ();
+                fn #visit_node(
+                    #self_param,
+                    node: #root_module::#node_ref_name<#root_module::#node_name>
+                ) -> Result<Self::#node_ret, Self::Error> {
+                    #impl_body
+                }
+            };
+        }
+    });
+
+    // @@Todo: hiding variant
+    // let all_nodes = tree_def
+    //     .nodes
+    //     .keys()
+    //     .map(|node_name| {
+    //         let node_name = format_ident!("{}",
+    // node_name.to_string().to_case(Case::Pascal));         quote! {
+    // #node_name, }     })
+    //     .collect::<TokenStream>();
+
+    // For each node name given, emit a default impl by calling the appropriate `@@`
+    // node case.
+    let result = quote! {
+        #[macro_export]
+        macro_rules! #default_impl_name {
+            ($($node:ident),* $(,)?) => {
+                $(
+                    #default_impl_name!(@@ $node);
+                )*
+            };
+            #(#default_impl_macro_cases)*
+            (@@ $node:ident) => {
+                compile_error!(concat!(
+                    "No such node type `",
+                    stringify!($node),
+                    "`",
+                ));
+            };
+        }
+    };
+
+    Ok(result)
+}
+
 /// Emit the tree definition as Rust syntax.
 pub(crate) fn emit_tree(tree_def: &TreeDef) -> Result<TokenStream, syn::Error> {
-    Ok([
-        emit_other_items(tree_def),
-        emit_node_defs(tree_def),
-        // Generate mut and non-mut versions of the walker and visitor
-        emit_visitor(tree_def, true, true),
-        emit_visitor(tree_def, false, true),
-        emit_visitor(tree_def, false, false),
-        emit_walker(tree_def, true, true)?,
-        emit_walker(tree_def, false, true)?,
-        emit_walker(tree_def, false, false)?,
-    ]
-    .into_iter()
-    .collect())
+    let mutability_combinations = [(true, true), (false, true), (false, false)];
```

Comment:
interesting that there is no case for `(true, false)`

Reply:
It wasn't used by anything, I guess I can add it but `(true, false)` is almost the same as `(true, true)` because most of the time when mutating ast nodes you have to mutate self as well in some way.

Reply:
yeah, no changes are required, just thought it was interesting that it wasn't a use case.